service/src/main/java/bio/terra/workspace/service/admin/flights/cloudcontexts/gcp/GcpIamCustomRolePatchStep.java

```diff
@@ -0,0 +1,185 @@
+import bio.terra.workspace.service.resource.controlled.cloud.gcp.CustomGcpIamRoleMapping;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.services.iam.v1.model.CreateRoleRequest;
+import com.google.api.services.iam.v1.model.Role;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.UUID;
+import org.apache.commons.collections.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Step to update IAM custom project and resource roles in GCP project. */
+public class GcpIamCustomRolePatchStep implements Step {
+  private final Logger logger = LoggerFactory.getLogger(GcpIamCustomRolePatchStep.class);
+
+  private final IamCow iamCow;
+  private final UUID workspaceId;
+  private final String projectId;
+
+  private final HashSet<CustomGcpIamRole> customGcpIamRoles = new HashSet<>();
+
+  public GcpIamCustomRolePatchStep(IamCow iamCow, UUID workspaceId, String projectId) {
+    this.iamCow = iamCow;
+    this.workspaceId = workspaceId;
+    this.projectId = projectId;
+    customGcpIamRoles.addAll(CUSTOM_GCP_IAM_ROLES);
+    customGcpIamRoles.addAll(CustomGcpIamRoleMapping.CUSTOM_GCP_RESOURCE_IAM_ROLES.values());
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    boolean workspaceUpdated = false;
+    for (CustomGcpIamRole customGcpIamRole : customGcpIamRoles) {
+      Role originalRole = getCustomRole(customGcpIamRole, projectId);
+      if (originalRole == null) {
+        createCustomRole(customGcpIamRole);
+        workspaceUpdated = true;
+      } else {
+        if (!CollectionUtils.isEqualCollection(
+            originalRole.getIncludedPermissions(), customGcpIamRole.getIncludedPermissions())) {
+          updateCustomRole(customGcpIamRole);
+          workspaceUpdated = true;
+        }
+      }
+    }
+    if (workspaceUpdated) {
+      FlightMap workingMap = flightContext.getWorkingMap();
+      HashSet<String> updatedWorkspaces =
+          workingMap.get(UPDATED_WORKSPACES, new TypeReference<>() {});
+      updatedWorkspaces.add(workspaceId.toString());
+      workingMap.put(UPDATED_WORKSPACES, updatedWorkspaces);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  /**
+   * Utility for creating custom roles in GCP from WSM's CustomGcpIamRole objects. These roles will
+   * be defined at the project level specified by projectId.
+   */
+  private void updateCustomRole(CustomGcpIamRole customRole) throws RetryException {
+    try {
+      // Only assigned field will be updated.
+      Role role = new Role().setIncludedPermissions(customRole.getIncludedPermissions());
+      iamCow
+          .projects()
+          .roles()
+          .patch(customRole.getFullyQualifiedRoleName(projectId), role)
+          .execute();
+      logger.debug(
```

Comment:
logger.info, same for rest of file

I think we're not configured to show debug logs in GCP Logging

Reply:
why are we configured to not show debug level?

i feel this is too noisy and i think debug is more like an on-demand logging. especially if there are a lot of workspaces

Reply:
The whole point of dry run is to sanity check these logs. So these logs need to be visible in GCP logging. (Unless there's some other way to confirm dry run, that I missed?)

Running this action is not common.

Reply:
I changed to Debug, I don't see any WSM debug logs in past 7 days

![image](https://user-images.githubusercontent.com/10929390/197602514-32daca76-5873-4dba-b2e1-c5e2b46f63c3.png)

https://github.com/DataBiosphere/terra-workspace-manager/blob/main/DEVELOPMENT.md#controlling-log-level says "By default logging at INFO level and above are displayed"

Reply:
odd to disable it. i usually find debug log uesful as info starts becoming spammy as the code base grows.

anyway, i will change them to info